lib/utils/compose.coffee

```diff
@@ -0,0 +1,732 @@
+
+	through { objectMode: true }, (str, enc, cb) ->
+		return cb(null, str) if not str?
+
+		if inline
+			return cb(null, { status: str })
+
+		if /^Successfully tagged /.test(str)
+			progress = undefined
+		else
+			if (match = stepRegex.exec(str))
+				step = match[1]
+				numSteps = match[2]
+			if step?
+				str = "Step #{step}/#{numSteps}: #{str}"
+				progress = parseInt(step, 10) * 100 // parseInt(numSteps, 10)
+
+		cb(null, { status: str, progress })
+
+pullProgressAdapter = (outStream) ->
+	return (e) ->
+		{ status, id, percentage, error, errorDetail } = e
+		if status?
+			status = status.replace(/^Status: /, '')
+		if id?
+			status = "#{id}: #{status}"
+		if percentage is 100
+			percentage = undefined
+		outStream.write({
+			status,
+			progress: percentage,
+			error: errorDetail?.message ? error
+		})
+
+createSpinner = ->
+	chars = '|/-\\'
+	index = 0
+	-> chars[(index++) % chars.length]
+
+class BuildProgressUI
+	constructor: (tty, descriptors) ->
+		_ = require('lodash')
+		chalk = require('chalk')
+		through = require('through2')
+
+		services = _.map(descriptors, 'serviceName')
+		prefix = chalk.blue('[Build]') + '   '
+		eventHandler = @_handleEvent
+
+		streams = _(services).map (service) ->
+			stream = through.obj (event, _enc, cb) ->
+				eventHandler(service, event)
+				cb()
+			stream.pipe(tty.stream)
+			[ service, stream ]
+		.fromPairs()
+		.value()
+
+		@_tty = tty
+		@_serviceToDataMap = {}
+		@_services = services
+
+		offset = 10 # account for escape sequences inserted for colouring
+		@_prefixWidth = offset + prefix.length + _.max(_.map(services, 'length'))
+		@_prefix = prefix
+
+		# these are to handle window wrapping
+		@_maxLineWidth = null
+		@_lineWidths = []
+
+		@_startTime = null
+		@_ended = false
+		@_cancelled = false
+		@_spinner = createSpinner()
+
+		@streams = streams
+
+	_handleEvent: (service, event) =>
+		@_serviceToDataMap[service] = event
+
+	_handleInterrupt: =>
+		@_cancelled = true
+		@end()
+		process.exit(130) # 128 + SIGINT
+
+	start: =>
+		process.on('SIGINT', @_handleInterrupt)
+		@_tty.hideCursor()
+		@_services.forEach (service) =>
+			@streams[service].write({ status: 'Preparing...' })
+		@_tid = setInterval(@_display, 1000 / 10)
+		@_startTime = Date.now()
+
+	end: (summary = null) =>
+		return if @_ended
+		@_ended = true
+		process.removeListener('SIGINT', @_handleInterrupt)
+		clearInterval(@_tid)
+
+		@_clear()
+		@_renderStatus(true)
+		@_renderSummary(summary ? @_getServiceSummary())
+		@_tty.showCursor()
+
+	_display: =>
+		@_clear()
+		@_renderStatus()
+		@_renderSummary(@_getServiceSummary())
+		@_tty.cursorUp(@_services.length + 1) # for status line
+
+	_clear: ->
+		@_tty.deleteToEnd()
+		@_maxLineWidth = @_tty.currentWindowSize().width
+
+	_getServiceSummary: ->
+		_ = require('lodash')
+
+		services = @_services
+		serviceToDataMap = @_serviceToDataMap
+
+		_(services).map (service) ->
+			{ status, progress, error } = serviceToDataMap[service] ? {}
```

Comment:
With this, we're only showing the latest line from the docker build. For large `RUN` steps, that's ends up being `Step 3/6:  ---> Running in adebb1ebfc2a`, not the actual command being run. That's not useful, and it's going to happen all the time, for all the most expensive steps in people's dockerfiles. You do get useful output for the initial `FROM` and `COPY` etc, but I still think we need to improve this.

Could we easily show all the lines from the current step, not just literally the last line? It'd require us to put the status on a separate line below, but that's actually a bit clearer imo. Might be some tricky UX with the output changing in height though... Alternatively, we could special case `RUN` steps to provide better output? Need to think about how to handle long multiline RUN commands though...

Reply:
I think it's good enough for now and there's also the option to inline everything. I'd save improvements for subsequent PRs.